lib/src/editor.dart

```diff
@@ -0,0 +1,499 @@
+import 'dart:collection' show UnmodifiableListView;
+
+import 'package:meta/meta.dart';
+import 'package:yaml/yaml.dart';
+
+import 'equality.dart';
+import 'list_mutations.dart';
+import 'map_mutations.dart';
+import 'source_edit.dart';
+import 'utils.dart';
+import 'wrap.dart';
+
+/// An interface for modififying [YAML][1] documents while preserving comments and
+/// whitespaces.
+///
+/// Most modification methods require the user to pass in an Iterable<Object> path that
+/// holds the keys/indices to navigate to the element. Key equality is performed via
+/// `package:yaml`'s [YamlMap]'s key equality.
+///
+/// For example,
+///
+/// ```yaml
+/// a: 1
+/// b: 2
+/// c:
+///   - 3
+///   - 4
+///   - {e: 5, f: [6, 7]}
+/// ```
+///
+/// To get to `7`, our path will be `['c', 2, 'f', 1]`. The path for the base object is the
+/// empty array `[]`. All modification methods will return a [ArgumentError] if the path
+/// provided is invalid. Note also that that the order of elements in the path is important,
+/// and it should be arranged in order of calling, with the first element being the first
+/// key or index to be called.
+///
+/// To dump the YAML after all the modifications have been completed, simply call [toString()].
+///
+/// [1]: https://yaml.org/
+@sealed
+abstract class YamlEditor {
+  /// Returns a list of [SourceEdit]s describing the modifications performed on the
+  /// YAML string thus far.
+  UnmodifiableListView<SourceEdit> get edits;
+
+  factory YamlEditor(String yaml) => YamlStringEditor(yaml);
+
+  /// Parses the document to return [YamlNode] currently present at [path]. If no [YamlNode]s exist
+  /// at [path], [parseAt] will return a [YamlNode]-wrapped [orElse] if it is defined, or throw an
+  /// [ArgumentError] otherwise.
+  ///
+  /// Note that this behavior deviates slightly from standard [Map] behavior, which returns `null`
+  /// instead if the key is not present.
```

Comment:
I think it would better to say:
```suggestion
  /// To get `null` when [path] does not point to a value in the [YamlNode]-tree, simply pass `orElse: null`.
  /// ```dart
  /// final myYamlEditor('{"key": "value"}');
  /// final value = myYamlEditor.valueAt(['invalid', 'path'], orElse: null);
  /// print(value) // null
  /// ```
```